### PR TITLE
create server group in deck

### PR DESCRIPTION
**Pull request by ashleykasim**

### CODE REVIEW

File: app/scripts/modules/openstack/serverGroup/configure/wizard/basicSettings.html [hunk 1]
```diff
@@ -0,0 +1,96 @@
+<div class="container-fluid form-horizontal" ng-controller="openstackServerGroupBasicSettingsCtrl as basicSettingsCtrl">
+  <div class="modal-body">
+    <ng-form name="basicSettings">
+      <div class="form-group">
+        <div class="col-md-3 sm-label-right">
+          Account
+        </div>
+        <div class="col-md-7">
+          <account-select-field required read-only="command.viewState.readOnlyFields.credentials" component="command" field="credentials" accounts="command.backingData.accounts"
+                                provider="'openstacke'"></account-select-field>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/basicSettings.html]: s/openstacke/openstack

File: app/scripts/modules/openstack/serverGroup/configure/wizard/capacity.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<form ng-controller="openstackServerGroupCapacityCtrl" class="container-fluid form-horizontal" name="form" novalidate>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/capacity.html]: I think this file can be omitted - doesn't look like it's referenced anywhere and it looks like the older style of modal.

(Might be worth removing it from whichever provider this PR is based on (kubernetes, I think), too)

File: app/scripts/modules/openstack/serverGroup/configure/wizard/deployInitializer.controller.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+let angular = require('angular');
+
+module.exports = angular.module('spinnaker.opensatck.serverGroup.configure.deployInitialization.controller', [
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/deployInitializer.controller.js]: s/opensatck/openstack

File: app/scripts/modules/openstack/serverGroup/configure/wizard/securityGroups.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
+<form ng-controller="openstackServerGroupSecurityGroupsCtrl as serverGroupSecurityGroupsController" class="container-fluid form-horizontal" name="form" novalidate>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/securityGroups.html]: Also appears to be unused - can likely be removed

**Comment by anotherchrisberry** [comment]
There are a number of files that can be pruned. Otherwise, it looks good on the first pass. I'll take a closer look this afternoon or this evening and try to get better comments in place. If you have time to trim out unused files, that would definitely make it easier to review!

**Comment by ashleykasim** [comment]
@anotherchrisberry fixed my spelling, made a pass through to try to trim out unneeded files, they were indeed coming from the Kubernetes example

**Comment by varikin** [comment]
Do we not need the copyright notices on the JS files like we do on Java/Groovy?

File: app/scripts/modules/openstack/image/image.reader.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        .then(function(results) {
+          return results;
+        })
+        .catch(function() {
+          return [];
+        });
+    }
+
+    function getImage(amiName, region, credentials) {
+      return Restangular.all('images').one(credentials).one(region).all(amiName).getList({provider: 'openstack'}).then(function(results) {
```
Comment (varikin) [inline app/scripts/modules/openstack/image/image.reader.js]: Any reason `getImage` assumes provider of openstack, but `findImages` requires `provider: 'openstack'` to be passed in?

File: app/scripts/modules/openstack/serverGroup/configure/serverGroup.configure.openstack.module.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+module.exports = angular.module('spinnaker.openstack.serverGroup.configure', [
+  require('../../../core/account/account.module.js'),
+  require('./wizard/deployInitializer.controller.js'),
+  require('./ServerGroupCommandBuilder.js'),
+  require('./serverGroupConfiguration.service.js'),
+  require('./wizard/ServerGroupBasicSettings.controller.js'),
+  require('./wizard/Capacity.controller.js'),
+  require('./wizard/Clone.controller.js'),
+  require('./wizard/LoadBalancers.controller.js'),
+  require('./wizard/templateSelection.controller.js')
```
Comment (varikin) [inline app/scripts/modules/openstack/serverGroup/configure/serverGroup.configure.openstack.module.js]: Should the capitalization of the files be consistent e.g. `templateSelection.controller.js` and `Clone.controller.js`

File: app/scripts/modules/openstack/image/image.reader.js (hunk 5 above)
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/image/image.reader.js]: It's inconsistent within this particular service, but consistent with the other `*ImageReader` service APIs. Not that it's "good" - it just hadn't been noticed or called out or fixed before.

I'm okay with it as-is. It might be nice (after this PR gets in) to remove the generic `imageReader` in the core module and refactor all the `*ImageReader` calls so that the provider doesn't need to be passed in.

File: app/scripts/modules/openstack/serverGroup/configure/serverGroup.configure.openstack.module.js (hunk 6 above)
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/serverGroup.configure.openstack.module.js]: This is consistent with Kubernetes, where @lwander inexplicably decided to start upper camel casing files and I inexplicably didn't say anything.

Lower camel case is preferred for file names. That said, getting git to accept file name changes where only the case changes can be a nightmare - I'm okay with this, with the caveat that someone might go in and change them (and k8s) to be consistent - very happy to accept a PR around that!

File: app/scripts/modules/openstack/serverGroup/configure/wizard/SecurityGroups.controller.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
+'use strict';
+
+let angular = require('angular');
+
+module.exports = angular.module('spinnaker.openstack.serverGroup.configure.securityGroups', [])
+  .controller('openstackServerGroupSecurityGroupsCtrl', function(modalWizardService) {
+    modalWizardService.getWizard().markClean('security-groups');
+    modalWizardService.getWizard().markComplete('security-groups');
+  });
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/SecurityGroups.controller.js]: unused file

File: app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      <v2-modal-wizard ng-show="state.loaded && !state.requiresTemplateSelection" heading="{{title}}">
+        <v2-wizard-page key="basic-settings" label="Basic Settings" hide-subheading="true">
+          <ng-include src="pages.basicSettings"></ng-include>
+        </v2-wizard-page>
+        <v2-wizard-page key="load-balancers" label="Load Balancers" mark-complete-on-view="false">
+          <ng-include src="pages.loadBalancers"></ng-include>
+        </v2-wizard-page>
+        <v2-wizard-page key="security-groups" label="Security Groups" mark-complete-on-view="false" done="true">
+          <ng-include src="pages.securityGroups"></ng-include>
+        </v2-wizard-page>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html]: Seems like this can be removed

File: app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        <v2-wizard-page key="capacity" label="Capacity" done="true">
+          <ng-include src="pages.capacity"></ng-include>
+        </v2-wizard-page>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html]: Can be removed

File: app/scripts/modules/openstack/serverGroup/serverGroup.transformer.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        var ports = '' + command.resources.ports;
+        command.resources.ports = ports.split(/\s*,\s*/);
+      }
+      if (command.securityGroups) {
+        var securityGroups = '' + command.securityGroups;
+        command.securityGroups = securityGroups.split(/\s*,\s*/);
+      }
+      if (command.securityGroups === '') {
+        delete command.securityGroups;
+      }
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/serverGroup.transformer.js]: Should be able to remove all the security group stuff (L26-32)

File: app/scripts/modules/openstack/serverGroup/serverGroup.transformer.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      if (command.resources.allocateIpAddress === true) {
+        delete command.resources.ports;
+      }
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/serverGroup.transformer.js]: Does this apply to Openstack?

File: app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        <div class="form-group" ng-if="command.viewState.customTemplateMessage">
+          <p class="col-md-10 col-md-offset-1" ng-bind-html="command.viewState.customTemplateMessage"></p>
+        </div>
+        <div class="form-group" ng-if="command.viewState.template.serverGroup" style="margin-top: 20px">
+          <div class="col-md-10 col-md-offset-1 well">
+            These fields <strong>will be</strong> copied over from the most recent server group:
+            <ul>
+              <li>account, namespace, cluster name (stack, details)</li>
+              <li>load balancers</li>
+              <li>security groups</li>
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html]: remove this line (I'm assuming security groups don't apply to Openstack here - disregard my comment if that's not the case)

File: app/scripts/modules/openstack/serverGroup/configure/wizard/loadBalancers.html [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        <ui-select multiple ng-model="command.loadBalancers" class="form-control input-sm">
+          <ui-select-match>{{$item}}</ui-select-match>
+          <ui-select-choices repeat="loadBalancer in command.backingData.filtered.loadBalancers | filter: $select.search">
+            <span ng-bind-html="loadBalancer | highlight: $select.search"></span>
+          </ui-select-choices>
+        </ui-select>
+      </div>
+    </div>
+  </div>
+  <div class="modal-footer">
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/loadBalancers.html]: This footer should go away.

File: app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html (hunk 12 above)
Comment (varikin) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html]: Yes, Openstack has security groups.

File: app/scripts/modules/openstack/serverGroup/configure/serverGroupConfiguration.service.js [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+    return {
+      configureUpdateCommand: configureUpdateCommand,
+      configureCommand: configureCommand,
+      configureImages: configureImages,
+      configureSecurityGroupOptions: configureSecurityGroupOptions,
+      configureLoadBalancerOptions: configureLoadBalancerOptions,
+      refreshLoadBalancers: refreshLoadBalancers,
+      refreshSecurityGroups: refreshSecurityGroups,
+      getRegionalSecurityGroups: getRegionalSecurityGroups,
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/serverGroupConfiguration.service.js]: Better to remove all the security group code (these two methods + `configureSecurityGroupOptions`) from this service - should knock the file down by 50 lines or so.

File: app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+      return $q.all({
+        images: imageLoader,
+      }).then(function (backingData) {
+        return {
+          application: application.name,
+          credentials: defaultCredentials,
+          region: defaultRegion,
+          images: backingData.images,
+          loadBalancers: [],
+          securityGroups: [],
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js]: remove

File: app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+          strategy: '',
+          sku: {
+            capacity: 1,
+          },
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js]: is `sku` applicable to Openstack?

File: app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+          viewState: {
+            instanceProfile: 'custom',
+            allImageSelection: null,
+            useAllImageSelection: false,
+            useSimpleCapacity: true,
+            usePreferredZones: true,
+            mode: defaults.mode || 'create',
+            disableStrategySelection: true,
+            loadBalancersConfigured: false,
+            securityGroupsConfigured: false,
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/ServerGroupCommandBuilder.js]: remove

File: app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html (hunk 12 above)
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html]: Security groups are not being surfaced in the wizard here - is that something you folks want to tackle later or as part of this PR?

Comment (ashleykasim) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/templateSelection.html]: That's going to happen later, at the time I was working on this code, the back end code for security groups wasn't merged. We will tackle security groups separately.

File: app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html (hunk 8 above)
Comment (ashleykasim) [inline app/scripts/modules/openstack/serverGroup/configure/wizard/serverGroupWizard.html]: Security groups will be needed in Openstack, so I'd like to leave this here to be implemented in another PR

**Comment by anotherchrisberry** [comment]
@ashleykasim I don't think we need to include license headers in every file. There's a LICENSE.txt file in the root directory that covers the entire repository, from what I've been told.
